57-monkfish/technical_specification.md

```diff
@@ -0,0 +1,99 @@
+# State Management Service (Monkfish)
+**Epic Type:** Implementation Epic
+
+Epic planning and implementation follow the
+[Epic Planning and Marathon SOP](https://docs.ghga-dev.de/main/sops/sop001_epic_planning.html).
+
+## Scope
+### Outline:
+The [Archive Test Bed](https://github.com/ghga-de/archive-test-bed) needs a way to
+consistently perform arbitrary CRUD operations
+in MongoDB that are outside the scope of the component services, but also unwieldy
+to manage in piecemeal code adaptations. Namely, the needs are:
+- setting a predefined state before running the tests (empty or prepopulate databases)
+- examining the state after or in between tests (whitebox testing)
+
+Right now, there's not a clean way to achieve either. 
+A dedicated service can offer a simple solution to these problems and allow test bed 
+processes to programmatically seed, reset, modify and examine databases through a single API.
+
+This service should **never** be deployed to production. It is *only* intended for use
+in the Testing and Staging environments, where there is no access to real data.  
+Despite this, there should be a way to restrict which databases and collections can be
+accessed with this service through configuration, and a simple API key (set in config)
+that can be used to authenticate requests.
+
+### Included/Required:
+The implementation of the SMS should include a token-secured RESTful API that interacts
+with MongoDB. There should be configuration to control access to databases and collections,
+as well as an API Key for authentication.
+
+Access control for database/collection resources should occur on a whitelist basis, with
+further specification for the individual permissions granted at the collection level
+for create, read, update, and delete.
+
+Since database names in the testing environment feature a prefix for isolation, the
+config should have a `db_prefix` value. This will be prefixed to all supplied database
+names, meaning the full name doesn't have to be supplied by the user in configuration,
+making the configuration more readable.
```

Comment:
Not only for the config, the `db_prefix` should also be silently added to all db names specified as path parameters.